src/python/pants/backend/docker/utils.py

```diff
@@ -47,3 +51,120 @@ def from_strings(cls: type[_T], *strings: str, duplicates_must_match: bool = Fal
+
+        Walks up the directory tree in case there is a recursive flag on one of the parent
+        directories.
+        """
+        if dirname is None:
+            dirname = os.path.dirname(path)
+        refs = referenced.get(dirname, set())
+        if isinstance(refs, bool):
+            return refs
+        if path in refs:
+            return True
+        parentdir = os.path.dirname(dirname)
+        if parentdir:
+            return is_referenced(path, parentdir)
+        return False
+
+    def get_unreferenced(files: Sequence[str] = (), dirs: Sequence[str] = ()) -> Iterator[str]:
+        unreferenced_files = tuple(path for path in files if not is_referenced(path))
+        yield from unreferenced_files
+        for path in dirs:
+            if not any(filename.startswith(path + "/") for filename in unreferenced_files):
+                # Skip paths where we don't have any unreferenced files any longer.
+                continue
+            if not is_referenced(path, path):
+                yield path
+
+    def get_matches(path: str) -> tuple[str, ...]:
+        is_pattern = any(all(c in path for c in cs) for cs in ["*", "?", "[]"])
+        if not is_pattern:
+            return (path,) if path in actual_paths else ()
+        #
+        # NOTICE: There is a slight difference in the pattern syntax used for the Dockerfile `COPY`
+        # instruction, than what is implmented by the `fnmatch` function in Python.
+        # https://docs.docker.com/engine/reference/builder/#copy which is implemented using
+        # https://golang.org/pkg/path/filepath#Match compared to
+        # https://docs.python.org/3/library/fnmatch.html#fnmatch.fnmatch
+        #
+        # For best experience when using globs, this should be addressed, but for now, I'll settle
+        # for a "close enough" approximation to get this moving.
+        return tuple(p for p in actual_paths if fnmatch(p, path))
+
+    # Go over exact matches first, so we don't target them as possible matches for renames.
+    unmatched_paths = set()
+    for path in tentative_paths:
+        matches = get_matches(path)
+        for match in matches:
+            reference(match)
+        if not matches:
+            unmatched_paths.add(path)
+
+    # List unknown files, possibly with a rename suggestion.
+    for path in sorted(unmatched_paths):
+        for suggestion in difflib.get_close_matches(path, actual_paths, n=1, cutoff=0.1):
+            # Suggest rename to match what files there are.
+            reference(suggestion)
+            yield path, suggestion
+            break
+        else:
+            # No match for this path.
+            yield path, ""
+
+    # List unused files.
+    for path in sorted(get_unreferenced(actual_files)):
+        yield "", path
+
+
+def format_rename_suggestion(src_path: str, dst_path: str, *, colors: bool) -> str:
+    """Given two paths, formats a line showing what to change in `src_path` to get to `dst_path`."""
```

Comment:
The example showing the inline edits to get from one filename to another is neat, but maybe not as practical as rendering the entire suggested alternative (as that's easier to copy-paste to apply a fix). Might be worth rendering the entire alternative?

Reply:
Right.. copy paste. Yeah could be.. there's a ratio limiting the number of changes presented. If you have a single character typo, it's much easier to spot with a `{ i => j }`, but true enough, that's not suitable for copy paste.. hmm.

Reply:
OK, I've put this behind an advanced docker option. I'd like some more opinions on what's most useful (also for myself to find out what I prefer). If we find one way superior to the other, I suggest we can drop the option then. :)

Reply:
I recommend not including the option (options fatigue) and always use the "atomic" mode. Imo, it's much easier to read and also copy-and-paste.

I was wondering what the motivation for non-atomic mode is? Piping to sed? Showing a minimal diff?

(Really nifty engineering though!)

Reply:
> I was wondering what the motivation for non-atomic mode is? Piping to sed? Showing a minimal diff?

It's not uncommon for me to stare at a message and fail to see what is wrong with it. As such, I tend to spend more time than needed trying to understand/see what it is that is wrong. The non-atomic mode is very clear where the issue is, for small typos, and it is quicker to fix the typo, than what it is to copy and paste the whole thing is.

However when the diff is more than a small typo, I think the atomic mode serves better.. 

As such, I'd like to use this feature live for a while, before locking down the decision hard, but I think you may be right here, that also for small typos it would be ok with atomic mode as it gives both current and new value together. (and seeing this go into 2.10.x, I can rid this A/B option before 2.10 rc0)

Reply:
Okay. How about we always use atomic mode for now, and then add this option in a followup if we get feedback for the need of non-atomic? (That feedback can be from you yourself, of course)